packages/front-end/components/OpenVisualEditorLink.tsx

```diff
@@ -45,49 +61,81 @@ const OpenVisualEditorLink: FC<{
     });
 
     return url;
-  }, [visualEditorUrl, id, changeIndex, apiHost]);
+  }, [apiHost, changeIndex, id, visualEditorUrl]);
 
-  const navigate = useCallback(() => {
-    track("Open visual editor", {
-      source: "visual-editor-ui",
-      status: "success",
+  const getVisualEditorKey = useCallback(async () => {
+    const res = await apiCall<{ key: string }>("/visual-editor/key", {
+      method: "GET",
     });
-    window.location.href = url;
+    return res.key;
+  }, [apiCall]);
+
+  const navigate = useCallback(async () => {
+    setShowExtensionDialog(false);
+
+    setIsNavigating(true);
+
+    const key = await getVisualEditorKey();
+
+    window.postMessage(
+      {
+        type: "GB_REQUEST_OPEN_VISUAL_EDITOR",
+        data: key,
+      },
+      window.location.origin
+    );
```

Comment:
If someone has the extension installed, but it's an old version, it looks like they will be stuck.  The extension dialog won't show, so there's no way to click the "bypass" link.  It will just stay "navigating" forever.  Can we add a simple `setTimeout` here.  If we haven't received a response from the extension within a short period, show a new dialog with a link letting them bypass.

This will happen to 100% of users when we merge this PR since the extension update won't be released yet.

Reply:
Also, there's no error handling here if the async calls like `getVisualEditorKey` fail.  We should probably catch those and do the same thing - set navigating back to false and show a dialog letting them proceed.  We would also need to refactor this a bit since currently, clicking the bypass link re-runs this entire navigate function, including the API calls, so if those are failing it will be a loop.